torch/distributed/_spmd/api.py

```diff
@@ -306,6 +318,43 @@ def _fused_adam_decomp(
 }
 
 
+DEDUP_TARGETS: Set[torch._ops.OpOverload] = {
+    aten.all_reduce.default,
+    aten.wait_tensor.default,
+}
+
+
+def _dedup_collectives(gm: fx.GraphModule) -> fx.GraphModule:
+    # deduplicate collectives
+    node_to_node: Dict[fx.Node, fx.Node] = {}
+    args_to_node: Dict[Tuple[Any, ...], fx.Node] = {}
+
+    nodes_to_erase: List[fx.Node] = []
+
+    for node in gm.graph.nodes:
+        # replace all args with the results from the first unique comm op
+        args, spec = pytree.tree_flatten(node.args)
+        unique_args = [node_to_node.get(a, a) for a in args]
+        node.args = pytree.tree_unflatten(unique_args, spec)
```

Comment:
hmmm should applying `node.replace_uses_with` to duplicate nodes just work instead of manually updating the args?